File: src/dmsg_types/src/cose.rs
//! ICP formal signing, vetKD derivation, execution results and key provenance.
//!
//! Client requests go to the user home for authorization; execution grants are
//! restricted user-to-COSE contracts. Construct approvals with `dmsg_protocol`.
//! Business times are Unix milliseconds; costs are ICP cycles.
use crate::*;
use candid::{CandidType, Principal};
use serde::{Deserialize, Serialize};
use serde_bytes::{ByteArray, ByteBuf};

/// Only algorithms that can produce a formal signature.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum SigningAlgorithm {
    /// Ed25519 formal signatures (COSE algorithm -19).
    Ed25519,
    /// ECDSA over secp256k1 (COSE ES256K, -47).
    EcdsaSecp256k1,
}
impl From<SigningAlgorithm> for Algorithm {
    fn from(value: SigningAlgorithm) -> Self {
        match value {
            SigningAlgorithm::Ed25519 => Self::Ed25519,
            SigningAlgorithm::EcdsaSecp256k1 => Self::EcdsaSecp256k1,
        }
    }
}
/// Domain of a formal document signing key; derived from the content profile.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum SigningPurpose {
    /// Key domain for text statements.
    Statement,
    /// Key domain for SHA-256 document attestations.
    FileAttestation,
}
/// Formal signing purpose and algorithm, with generation fixed to 1 on conversion.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct SigningKey {
    /// Key usage domain; part of key derivation.
    pub purpose: SigningPurpose,
    /// Cryptographic algorithm selected for this key or operation.
    pub algorithm: SigningAlgorithm,
}
/// Typed selector separating formal signing keys from vetKD content roots.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum KeySelector {
    /// Purpose-specific formal signing key.
    Signing(SigningKey),
    /// Content-root derivation domain selected by generation.
    ContentRoot {
        /// Content-root generation to derive.
        generation: u64,
    },
}
impl From<KeySelector> for KeyRequest {
    fn from(value: KeySelector) -> Self {
        match value {
            KeySelector::Signing(key) => Self {
                purpose: match key.purpose {
                    SigningPurpose::Statement => KeyPurpose::Statement,
                    SigningPurpose::FileAttestation => KeyPurpose::FileAttestation,
                },
                algorithm: key.algorithm.into(),
                generation: 1,
            },
            KeySelector::ContentRoot { generation } => Self {
                purpose: KeyPurpose::ContentRoot,
                algorithm: Algorithm::VetKdBls12381,
                generation,
            },
        }
    }
}

/// Small reference obtained from the authenticated key descriptor.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct SigningKeyRef {
    /// Cryptographic algorithm selected for this key or operation.
    pub algorithm: SigningAlgorithm,
    /// Opaque COSE key identifier from an authenticated descriptor.
    pub kid: ByteBuf,
    /// RFC 9679 SHA-256 thumbprint of required public COSE key parameters.
    pub public_key_fingerprint: Hash,
}
/// Formal signing request submitted to the user canister.
/// Freeze the statement, authenticated key reference, origin and cycle limit
/// before computing the approval. The portable statement has no approval deadline.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct SignRequest {
    /// Stable 12-byte dMsg account identity; not a Principal or ledger account.
    pub account_id: AccountId,
    /// Algorithm, kid and thumbprint from the authenticated signing-key descriptor.
    pub key: SigningKeyRef,
    /// Prepared or parsed document claims and content.
    pub statement: Statement,
    /// Checked browser origin; part of device approval, not of the portable statement.
    pub origin: String,
    /// Maximum ICP cycles approved for this execution; not a token amount.
    pub max_cycles: u128,
    /// Device approval binding the complete request and replay context.
    pub approval: Approval,
}

/// Select the committed root or the candidate of a reserved root operation.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum RootTarget {
    /// Currently committed content root.
    Current {
        /// Committed root generation.
        generation: u64,
    },
    /// Candidate root associated with an active reservation.
    Candidate {
        /// Reserved next root generation.
        generation: u64,
        /// Operation ID of the root reservation.
        op_id: OpId,
    },
}
/// Request a vetKD root encrypted to a caller-generated transport public key.
/// The result is encrypted key material, not a plaintext vault key.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct DeriveRootRequest {
    /// Stable 12-byte dMsg account identity; not a Principal or ledger account.
    pub account_id: AccountId,
    /// Committed or reserved candidate root to derive.
    pub target: RootTarget,
    /// 48-byte compressed vetKD transport public key; protocol validation checks the point.
    pub transport_public_key: ByteArray<48>,
    /// Maximum ICP cycles approved for this execution; not a token amount.
    pub max_cycles: u128,
    /// Device approval binding the complete request and replay context.
    pub approval: Approval,
}
impl DeriveRootRequest {
    /// Convert the typed root target to an execution request without validating or authorizing it.
    pub fn into_execution(self) -> ExecuteRequest {
        let (generation, root_op_id) = match self.target {
            RootTarget::Current { generation } => (generation, None),
            RootTarget::Candidate { generation, op_id } => (generation, Some(op_id)),
        };
        ExecuteRequest {
            account_id: self.account_id,
            max_cycles: self.max_cycles,
            approval: self.approval,
            kind: ExecutionKind::Derive {
                generation,
                root_op_id,
                transport_key: self.transport_public_key.to_vec().into(),
            },
        }
    }
}

/// Domain separation for formal signatures and content-root derivation.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
pub enum KeyPurpose {
    /// Key domain for SHA-256 document attestations.
    FileAttestation,
    /// Key domain for text statements.
    Statement,
    /// Content-root derivation domain, selected by generation.
    ContentRoot,
}
/// Supported chain-key operations; vetKD derives keys and cannot sign documents.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum Algorithm {
    /// Ed25519 formal signatures (COSE algorithm -19).
    Ed25519,
    /// ECDSA over secp256k1 (COSE ES256K, -47).
    EcdsaSecp256k1,
    /// BLS12-381 vetKD encrypted key derivation; not a document-signature algorithm.
    VetKdBls12381,
}
/// Pinned ICP master-key configuration checked during COSE initialization.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct MasterKey {
    /// Cryptographic algorithm selected for this key or operation.
    pub algorithm: Algorithm,
    /// ICP management-canister master-key name.
    pub key_name: String,
    /// SHA-256 of raw management public-key bytes, not a COSE thumbprint.
    /// A zero pin is allowed only outside Production.
    pub expected_fingerprint: Hash,
}
/// Deployment configuration for the COSE executor.
/// Production validation requires fixed master names and nonzero fingerprints;
/// changing derivation inputs changes key identity.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct CoseInit {
    /// Fixed issuer URI prefix; ends in `/` or `:` with no query or fragment.
    pub issuer_namespace: String,
    /// Deployment domain used in validation and derivation.
    pub environment: Environment,
    /// Expected identity of this COSE executor.
    pub executing_canister: Principal,
    /// User canister authorized to supply execution grants.
    pub initial_home_user: Principal,
    /// Key derivation format version; current public protocol uses 2.
    pub derivation_version: u16,
    /// Configured algorithm/master-key pins.
    pub masters: Vec<MasterKey>,
    /// Maximum executions in a daily budget window.
    pub daily_executions: u32,
    /// Maximum ICP cycles in a daily budget window.
    pub daily_cycles: u128,
}

/// Readiness of configured chain-key public keys.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum Initialization {
    /// Required material has not been initialized.
    Uninitialized,
    /// Public-key initialization is in progress.
    Initializing,
    /// Required material is initialized and ready for use.
    Ready,
}
/// Public key provenance and derivation identity.
/// Authenticate its source before using it to bind a key to an account.
/// Key identity depends on the home canister and derivation configuration.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct KeyDescriptor {
    /// COSE thumbprint bytes for formal signing; domain-separated derivation
    /// identifier for vetKD. Treat as opaque.
    pub key_id: ByteBuf,
    /// Stable 12-byte dMsg account identity; not a Principal or ledger account.
    pub account_id: AccountId,
    /// Key usage domain; part of key derivation.
    pub purpose: KeyPurpose,
    /// Cryptographic algorithm selected for this key or operation.
    pub algorithm: Algorithm,
    /// Fixed COSE canister responsible for key derivation and execution.
    pub home_cose: Principal,
    /// ICP master-key name used for this derived key.
    pub master_key_name: String,
    /// Deployment domain used in validation and derivation.
    pub environment: Environment,
    /// Key derivation format version; current public protocol uses 2.
    pub derivation_version: u16,
    /// Generation included in key derivation; formal signing currently uses 1.
    pub key_generation: u64,
    /// Raw Ed25519 key (32 bytes), SEC1 secp256k1 key, or vetKD public key
    /// (96 bytes). This is not an encoded COSE_Key.
    pub public_key: ByteBuf,
    /// RFC 9679 SHA-256 thumbprint for signing keys; SHA-256 of raw public
    /// key bytes for vetKD.
    pub public_key_fingerprint: Hash,
}
/// Low-level key derivation selector. Prefer [`KeySelector`] when constructing requests.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct KeyRequest {
    /// Key usage domain; part of key derivation.
    pub purpose: KeyPurpose,
    /// Cryptographic algorithm selected for this key or operation.
    pub algorithm: Algorithm,
    /// Positive key generation; formal signing requires 1, content roots use
    /// the requested root generation.
    pub generation: u64,
}

/// Exact operation covered by device approval and execution authorization.
/// This low-level contract is not an unrestricted raw-signing API.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum ExecutionKind {
    /// Formal document signing with fixed key, prepared bytes and origin.
    Sign {
        /// Purpose, algorithm and generation of the formal signing key.
        key: KeyRequest,
        /// Exact canonical COSE Sig_structure bytes; not a raw document hash.
        to_be_signed: ByteBuf,
        /// RFC 9679 SHA-256 thumbprint of required public COSE key parameters.
        public_key_fingerprint: Hash,
        /// Checked browser origin bound by the device approval.
        origin: String,
    },
    /// Content-root derivation encrypted to a vetKD transport key.
    Derive {
        /// Generation of the selected root or recovery material.
        generation: u64,
        /// Reservation operation ID for a candidate root; None for the committed root.
        root_op_id: Option<OpId>,
        /// 48-byte compressed vetKD transport public key after request conversion.
        transport_key: ByteBuf,
    },
}
/// Operation submitted for account policy checks and device authorization.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct ExecuteRequest {
    /// Stable 12-byte dMsg account identity; not a Principal or ledger account.
    pub account_id: AccountId,
    /// Exact signing or root-derivation operation covered by approval.
    pub kind: ExecutionKind,
    /// Maximum ICP cycles approved for this execution; not a token amount.
    pub max_cycles: u128,
    /// Device approval binding the complete request and replay context.
    pub approval: Approval,
}

/// User-home authorization passed to the configured COSE home.
/// This is a restricted cross-canister contract, not a bearer token that any
/// caller may construct and redeem.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct ExecutionGrant {
    /// Stable 12-byte dMsg account identity; not a Principal or ledger account.
    pub account_id: AccountId,
    /// User canister authoritative for this account or deployment.
    pub home_user: Principal,
    /// Fixed COSE canister responsible for key derivation and execution.
    pub home_cose: Principal,
    /// Operation identity used to bind approval and reconcile retries.
    pub request_id: OpId,
    /// User-home execution sequence for replay protection.
    pub execution_sequence: u64,
    /// Account security revision used to invalidate stale approvals.
    pub security_epoch: u64,
    /// 32-byte device identifier, distinct from its signing public key.
    pub device_id: Hash,
    /// Device approval sequence consumed by this authorization.
    #[serde(default)]
    pub device_sequence: u64,
    /// Time authorization was accepted, in Unix milliseconds.
    pub approved_at: u64,
    /// Exclusive deadline in Unix milliseconds (`now < expires_at`).
    pub expires_at: u64,
    /// Authorized operation, identical to the approved request.
    pub kind: ExecutionKind,
    /// Maximum ICP cycles approved for this execution; not a token amount.
    pub max_cycles: u128,
}
/// Execution lifecycle without the result payload.
/// Unknown outcomes must be reconciled using the original request ID.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum ExecutionStatus {
    /// Authorization is committed; no completed result is yet available.
    Authorized,
    /// Execution has started and may be awaiting an external call.
    Executing,
    /// Execution completed with a retained output.
    Completed,
    /// Execution returned a known failure.
    Failed,
    /// Execution outcome is uncertain; reconcile using the same request ID.
    Unknown,
    /// Retained output is no longer available; this does not authorize replay.
    ResultExpired,
}
/// Query/retry view of an execution and its charged cycles.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct ExecutionResult {
    /// Operation identity used to bind approval and reconcile retries.
    pub request_id: OpId,
    /// Execution state with any available output or error.
    pub outcome: ExecutionOutcome,
    /// ICP cycles charged for this execution.
    pub charged_cycles: u128,
}
/// Execution lifecycle carrying either output or failure information.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum ExecutionOutcome {
    /// Authorization is committed; no completed result is yet available.
    Authorized,
    /// Execution has started and may be awaiting an external call.
    Executing,
    /// Execution completed with a retained output.
    Completed(Box<ExecutionOutput>),
    /// Execution returned a known failure.
    Failed(Error),
    /// Execution outcome is uncertain; reconcile using the same request ID.
    Unknown(Error),
    /// Retained output is no longer available; this does not authorize replay.
    ResultExpired,
}
/// Successful formal signature or encrypted vetKD derivation result.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum ExecutionOutput {
    /// Portable signed document and its key provenance.
    Signature {
        /// Portable COSE document and public key.
        artifact: SignedArtifact,
        /// Public descriptor of the key used to produce this output.
        key: KeyDescriptor,
    },
    /// Encrypted vetKD output and the derivation-key descriptor.
    EncryptedRootKey {
        /// vetKD result encrypted to the supplied transport public key.
        encrypted_key: ByteBuf,
        /// Public descriptor of the key used to produce this output.
        key: KeyDescriptor,
    },
}
impl ExecutionOutput {
    /// Borrow the COSE_Sign1 bytes or encrypted vetKD bytes, according to the variant.
    pub fn bytes(&self) -> &ByteBuf {
        match self {
            Self::Signature { artifact, .. } => &artifact.cose_sign1,
            Self::EncryptedRootKey { encrypted_key, .. } => encrypted_key,
        }
    }
    /// Borrow the public descriptor attached to either output variant.
    pub fn key(&self) -> &KeyDescriptor {
        match self {
            Self::Signature { key, .. } | Self::EncryptedRootKey { key, .. } => key,
        }
    }
}
impl ExecutionResult {
    /// Project the outcome to its payload-free lifecycle status.
    pub fn status(&self) -> ExecutionStatus {
        match self.outcome {
            ExecutionOutcome::Authorized => ExecutionStatus::Authorized,
            ExecutionOutcome::Executing => ExecutionStatus::Executing,
            ExecutionOutcome::Completed(_) => ExecutionStatus::Completed,
            ExecutionOutcome::Failed(_) => ExecutionStatus::Failed,
            ExecutionOutcome::Unknown(_) => ExecutionStatus::Unknown,
            ExecutionOutcome::ResultExpired => ExecutionStatus::ResultExpired,
        }
    }
    /// Whether this result is Completed, Failed or ResultExpired.
    /// Unknown is deliberately nonterminal and must be reconciled.
    pub fn is_terminal(&self) -> bool {
        matches!(
            self.outcome,
            ExecutionOutcome::Completed(_)
                | ExecutionOutcome::Failed(_)
                | ExecutionOutcome::ResultExpired
        )
    }
    /// Borrow completed output.
    ///
    /// # Errors
    /// Returns the stored failure/unknown error, ResultExpired for pruned output,
    /// or Pending while authorization/execution is in progress.
    pub fn output(&self) -> Result<&ExecutionOutput> {
        match &self.outcome {
            ExecutionOutcome::Completed(output) => Ok(output),
            ExecutionOutcome::Failed(e) | ExecutionOutcome::Unknown(e) => Err(e.clone()),
            ExecutionOutcome::ResultExpired => Err(Error::ResultExpired),
            _ => Err(Error::Pending),
        }
    }
}

/// COSE deployment configuration and master-key initialization diagnostics.
#[derive(CandidType, Serialize, Deserialize, Clone)]
pub struct KeyState {
    /// Deployment configuration used by the executor.
    pub config: CoseInit,
    /// Master public-key initialization state.
    pub initialization: Initialization,
    /// SHA-256 of raw initialized master public keys, in config.masters order.
    pub fingerprints: Vec<Hash>,
    /// Initialization diagnostic, when one is available.
    pub error: Option<String>,
}

/// Certified execution evidence binding an operation to the signed bytes and key.
/// Current schema is 1; the single leaf path is `b"execution/" || account_id || request_id`.
/// Authenticate the certificate and witness before matching these fields to an artifact.
/// This records dMsg authorization, not external application permissions or a TSA timestamp.
#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub struct ExecutionReceipt {
    /// Version of this public certified-leaf format, not the storage schema.
    pub schema: u16,
    /// Stable 12-byte dMsg account identity; not a Principal or ledger account.
    pub account_id: AccountId,
    /// Canonical absolute issuer URI identifying the signer.
    pub issuer: String,
    /// Operation identity used to bind approval and reconcile retries.
    pub request_id: OpId,
    /// 32-byte device identifier, distinct from its signing public key.
    pub device_id: Hash,
    /// Account security revision used to invalidate stale approvals.
    pub security_epoch: u64,
    /// Time authorization was accepted, in Unix milliseconds.
    pub approved_at: u64,
    /// Exclusive deadline in Unix milliseconds (`now < expires_at`).
    pub expires_at: u64,
    /// Checked browser origin bound by the device approval.
    pub origin: String,
    /// Maximum ICP cycles approved for this execution; not a token amount.
    pub max_cycles: u128,
    /// SHA-256 of the exact COSE Sig_structure bytes.
    pub to_be_signed_digest: Hash,
    /// RFC 9679 SHA-256 thumbprint of required public COSE key parameters.
    pub public_key_fingerprint: Hash,
    /// Lifecycle state recorded for the authorized execution.
    pub status: ExecutionStatus,
    /// SHA-256 of raw signature bytes, if available; differs from the CTT imprint.
    pub signature_digest: Option<Hash>,
}

File: src/dmsg_app/src/lib/services/cose.ts
import { encodeControl, decodeControl } from '../protocol/account'
import { actionToCandid, actionFromCandid } from '../protocol/app-action'
import { Principal } from '@icp-sdk/core/principal'
import type {
  _SERVICE as UserService,
  Approval,
  DeriveRootRequest,
  ExecutionResult,
  Statement,
  RootTarget,
  SignRequest,
  AppActionSignRequest,
  SigningAlgorithm
} from '../canisters/generated/user'
import type { _SERVICE as CoseService, KeySelector } from '../canisters/generated/cose'
import { digest, equal, hex } from '../protocol/codec'
import { DmsgError, ensure } from '../errors'
import { controlResult } from './account'
import {
  statementBytes,
  statementPurpose,
  verifyDocumentArtifact,
  type Algorithm,
  type DocumentStatement
} from '../protocol/statements'
export type { Algorithm } from '../protocol/statements'

export class ExecutionRejected extends DmsgError {}

export interface SigningKeyReference {
  algorithm: Algorithm
  kid: Uint8Array
  publicKeyFingerprint: Uint8Array
}

/** Obtain device/epoch/sequence from the authenticated account before preparing.
 * The origin must come from the extension's checked browser message, not a
 * third-party request field. A device signature cannot prove browser provenance.
 */
export interface ExecutionContext {
  homeUser: Principal
  accountId: Uint8Array
  issuer: string
  deviceId: Uint8Array
  securityEpoch: bigint
  sequence: bigint
  expiresAt: bigint
  maxCycles: bigint
}
type Operation =
  | { kind: 'sign'; request: SignRequest }
  | { kind: 'sign_app_action'; request: AppActionSignRequest }
  | { kind: 'derive_root'; request: DeriveRootRequest }
function cloneOperation(operation: Operation): Operation {
  // structuredClone strips Principal's prototype. Clone through the exact public
  // Candid schema so receiver identity survives approval, journaling and retry.
  return {
    kind: operation.kind,
    request: decodeControl(
      operation.kind,
      encodeControl(operation.kind, [operation.request])
    )[0]
  } as Operation
}
type UserExecution = Pick<
  UserService,
  'sign' | 'sign_app_action' | 'derive_root' | 'get_execution' | 'reconcile_execution'
>
export type DeviceSigner = (approvalDigest: Uint8Array) => Promise<Uint8Array>

function fixed(value: Uint8Array, size: number) {
  ensure(
    value instanceof Uint8Array && value.length === size,
    'INVALID_INPUT',
    `需要 ${size} 字节。`
  )
  ensure(
    value.some((byte) => byte !== 0),
    'INVALID_INPUT',
    '标识或公钥不能全为零。'
  )
  return Uint8Array.from(value)
}
function uint64(value: bigint) {
  ensure(
    typeof value === 'bigint' && value >= 0n && value <= 0xffffffffffffffffn,
    'INVALID_INPUT'
  )
  return value
}
function approval(context: ExecutionContext, now: number): Approval {
  const account = fixed(context.accountId, 12),
    device = fixed(context.deviceId, 32)
  const epoch = uint64(context.securityEpoch),
    sequence = uint64(context.sequence)
  const expiresAt = uint64(context.expiresAt)
  ensure(
    Number.isSafeInteger(now) && expiresAt > BigInt(now) && expiresAt <= BigInt(now) + 300000n,
    'EXPIRED',
    '批准有效期必须在未来 5 分钟内。'
  )
  ensure(context.maxCycles > 0n && context.maxCycles <= 100000000000n, 'QUOTA_EXCEEDED')
  ensure(
    !context.homeUser.isAnonymous() && context.homeUser.toUint8Array().length > 0,
    'INVALID_INPUT'
  )
  return {
    device_id: device,
    security_epoch: epoch,
    sequence,
    request_id: digest('dmsg/execution-request/v2', [account, epoch, device, sequence]),
    expires_at: expiresAt,
    signature: new Uint8Array()
  }
}
function toStatement(input: DocumentStatement): Statement {
  let content: Statement['content']
  switch (input.content.kind) {
    case 'app_action':
      content = { AppAction: actionToCandid(input.content.action) }
      break
    case 'text':
      content = { Text: input.content.text }
      break
    case 'digest':
      content = {
        Digest: {
          sha256: input.content.sha256,
          content_type:
            input.content.contentType === undefined ? [] : [input.content.contentType],
          location: input.content.location === undefined ? [] : [input.content.location]
        }
      }
      break
    case 'file_statement':
      content = {
        FileStatement: {
          text: input.content.text,
          sha256: input.content.sha256,
          content_type:
            input.content.contentType === undefined ? [] : [input.content.contentType],
          location: input.content.location === undefined ? [] : [input.content.location]
        }
      }
      break
  }
  return {
    issuer: input.issuer,
    subject: input.subject === undefined ? [] : [input.subject],
    issued_at: input.issuedAt === undefined ? [] : [input.issuedAt],
    content
  }
}
function fromStatement(s: Statement): DocumentStatement {
  return {
    issuer: s.issuer,
    subject: s.subject[0],
    issuedAt: s.issued_at[0],
    content:
      'AppAction' in s.content
        ? { kind: 'app_action', action: actionFromCandid(s.content.AppAction) }
        : 'Text' in s.content
          ? { kind: 'text', text: s.content.Text }
          : 'FileStatement' in s.content
            ? {
                kind: 'file_statement',
                text: s.content.FileStatement.text,
                sha256: Uint8Array.from(s.content.FileStatement.sha256),
                contentType: s.content.FileStatement.content_type[0],
                location: s.content.FileStatement.location[0]
              }
            : {
                kind: 'digest',
                sha256: Uint8Array.from(s.content.Digest.sha256),
                contentType: s.content.Digest.content_type[0],
                location: s.content.Digest.location[0]
              }
  }
}
export function signingRequest(request: SignRequest | AppActionSignRequest): SignRequest {
  return 'action' in request
    ? {
        ...request,
        origin: request.action.origin,
        statement: {
          issuer: request.issuer,
          subject: [],
          issued_at: [],
          content: { AppAction: request.action }
        }
      }
    : request
}
export function signBytes(input: SignRequest | AppActionSignRequest) {
  const request = signingRequest(input)
  return statementBytes(
    fromStatement(request.statement),
    Object.keys(request.key.algorithm)[0] as Algorithm,
    Uint8Array.from(request.key.kid)
  ).toBeSigned
}
function executionKind(operation: Operation): unknown {
  if (operation.kind !== 'derive_root') {
    const sign = signingRequest(operation.request)
    return {
      Sign: {
        key: {
          purpose: statementPurpose(fromStatement(sign.statement).content),
          algorithm: Object.keys(sign.key.algorithm)[0],
          generation: 1n
        },
        to_be_signed: signBytes(sign),
        public_key_fingerprint: Uint8Array.from(sign.key.public_key_fingerprint),
        origin: sign.origin
      }
    }
  }
  const root = operation.request
  const target = 'Current' in root.target ? root.target.Current : root.target.Candidate
  return {
    Derive: {
      generation: target.generation,
      root_op_id:
        'Candidate' in root.target ? Uint8Array.from(root.target.Candidate.op_id) : null,
      transport_key: Uint8Array.from(root.transport_public_key)
    }
  }
}
/** A snapshot of exactly what will be approved. Returned views are copies.
 * Persist the signed operation (and the root transport-secret reference) with
 * `onApproved` before dispatch when integrating an encrypted outbox.
 */
export class PreparedExecution {
  private readonly operation: Operation
  private readonly message: Uint8Array
  private submission?: Promise<ExecutionResult>

  constructor(operation: Operation, homeUser: Principal) {
    this.operation = cloneOperation(operation)
    const request = this.operation.request,
      a = request.approval
    this.message = digest('dmsg/device-approval/v2', [
      homeUser.toUint8Array(),
      Uint8Array.from(request.account_id),
      'dmsg/execute/v3',
      Uint8Array.from(a.device_id),
      a.security_epoch,
      a.sequence,
      Uint8Array.from(a.request_id),
      a.expires_at,
      digest('dmsg/execute/v3', [executionKind(this.operation), request.max_cycles])
    ])
  }
  get requestId() {
    return hex(Uint8Array.from(this.operation.request.approval.request_id))
  }
  get approvalMessage() {
    return Uint8Array.from(this.message)
  }
  get review() {
    return cloneOperation(this.operation)
  }
  get toBeSigned() {
    return this.operation.kind !== 'derive_root' ? signBytes(this.operation.request) : null
  }
  approveAndExecute(
    user: Pick<UserExecution, 'sign' | 'sign_app_action' | 'derive_root'>,
    signer: DeviceSigner,
    onApproved: (operation: Operation) => Promise<void> = async () => {}
  ): Promise<ExecutionResult> {
    this.submission ??= this.submit(user, signer, onApproved)
    return this.submission
  }
  private async submit(
    user: Pick<UserExecution, 'sign' | 'sign_app_action' | 'derive_root'>,
    signer: DeviceSigner,
    onApproved: (operation: Operation) => Promise<void>
  ) {
    const signed = cloneOperation(this.operation)
    signed.request.approval.signature = fixed(await signer(this.approvalMessage), 64)
    await onApproved(cloneOperation(signed))
    let response: Awaited<ReturnType<UserService['sign']>>
    try {
      response =
        signed.kind === 'sign'
          ? await user.sign(signed.request)
          : signed.kind === 'sign_app_action'
            ? await user.sign_app_action(signed.request)
            : await user.derive_root(signed.request)
    } catch {
      throw new DmsgError(
        'EXECUTION_UNKNOWN',
        `执行结果尚未确认，请按请求 ${this.requestId} 对账。`
      )
    }
    if ('Err' in response) {
      try {
        controlResult(response)
      } catch (error) {
        if (error instanceof DmsgError) throw new ExecutionRejected(error.code, error.message)
        throw error
      }
    }
    const result = controlResult(response)
    ensure(
      equal(
        Uint8Array.from(result.request_id),
        Uint8Array.from(signed.request.approval.request_id)
      ),
      'INTEGRITY_FAILED',
      '响应与原请求不一致。'
    )
    if (signed.kind !== 'derive_root' && 'Completed' in result.outcome) {
      const output = result.outcome.Completed
      ensure('Signature' in output, 'INTEGRITY_FAILED')
      const checked = verifyDocumentArtifact(output.Signature.artifact)
      ensure(
        equal(checked.toBeSigned, signBytes(signed.request)) &&
          equal(
            checked.keyFingerprint,
            Uint8Array.from(signed.request.key.public_key_fingerprint)
          ),
        'INTEGRITY_FAILED',
        '签名结果与批准的内容或密钥不符。'
      )
    }
    return result
  }
}

export function prepareSign(
  context: ExecutionContext,
  content: { origin: string; key: SigningKeyReference; statement: DocumentStatement },
  now = Date.now()
) {
  const { key, statement } = content
  ensure(statement.issuer === context.issuer, 'FORBIDDEN', '签署者必须与已认证账户一致。')
  statementBytes(statement, key.algorithm, key.kid)
  const origin = new URL(content.origin)
  ensure(
    content.origin.length <= 256 &&
      ((statement.content.kind === 'app_action' &&
        statement.content.action.origin === content.origin) ||
        (origin.protocol === 'https:' && origin.origin === content.origin) ||
        /^chrome-extension:\/\/[a-p]{32}$/.test(content.origin)),
    'INVALID_INPUT',
    '需要规范的应用 origin。'
  )
  if (statement.content.kind === 'app_action')
    return new PreparedExecution(
      {
        kind: 'sign_app_action',
        request: {
          account_id: fixed(context.accountId, 12),
          issuer: context.issuer,
          action: actionToCandid(statement.content.action),
          key: {
            algorithm: { [key.algorithm]: null } as SigningAlgorithm,
            kid: Uint8Array.from(key.kid),
            public_key_fingerprint: fixed(key.publicKeyFingerprint, 32)
          },
          max_cycles: context.maxCycles,
          approval: approval(context, now)
        }
      },
      context.homeUser
    )
  return new PreparedExecution(
    {
      kind: 'sign',
      request: {
        account_id: fixed(context.accountId, 12),
        key: {
          algorithm: { [key.algorithm]: null } as SigningAlgorithm,
          kid: Uint8Array.from(key.kid),
          public_key_fingerprint: fixed(key.publicKeyFingerprint, 32)
        },
        statement: toStatement(statement),
        origin: content.origin,
        max_cycles: context.maxCycles,
        approval: approval(context, now)
      }
    },
    context.homeUser
  )
}

export function prepareRootDerivation(
  context: ExecutionContext,
  target:
    | { kind: 'current'; generation: bigint }
    | { kind: 'candidate'; generation: bigint; opId: Uint8Array },
  transportPublicKey: Uint8Array,
  now = Date.now()
) {
  ensure(uint64(target.generation) > 0n, 'INVALID_INPUT')
  ensure(target.kind === 'current' || target.kind === 'candidate', 'INVALID_INPUT')
  const root: RootTarget =
    target.kind === 'current'
      ? { Current: { generation: target.generation } }
      : { Candidate: { generation: target.generation, op_id: fixed(target.opId, 32) } }
  return new PreparedExecution(
    {
      kind: 'derive_root',
      request: {
        account_id: fixed(context.accountId, 12),
        target: root,
        transport_public_key: fixed(transportPublicKey, 48),
        max_cycles: context.maxCycles,
        approval: approval(context, now)
      }
    },
    context.homeUser
  )
}

export async function signingKey(
  cose: Pick<CoseService, 'public_key'>,
  accountId: Uint8Array,
  purpose: 'statement' | 'file_attestation' | 'app_action',
  algorithm: Algorithm = 'Ed25519'
) {
  ensure(['Ed25519', 'EcdsaSecp256k1'].includes(algorithm), 'UNSUPPORTED_PROTOCOL')
  const key: KeySelector = {
    Signing: {
      purpose:
        purpose === 'app_action'
          ? { AppAction: null }
          : purpose === 'statement'
            ? { Statement: null }
            : { FileAttestation: null },
      algorithm: { [algorithm]: null } as SigningAlgorithm
    }
  }
  return controlResult(await cose.public_key(fixed(accountId, 12), key))
}
/** Reads the stored result of an approved execution, reconciling one still in
 * flight. Null means no execution is stored under this request ID. */
export async function recordedExecution(
  user: Pick<UserExecution, 'get_execution' | 'reconcile_execution'>,
  accountId: Uint8Array,
  requestId: Uint8Array
): Promise<ExecutionResult | null> {
  const account = fixed(accountId, 12),
    request = fixed(requestId, 32),
    response = await user.get_execution(account, request)
  if ('Err' in response) {
    ensure('ResultExpired' in response.Err, 'EXECUTION_UNKNOWN')
    return null
  }
  const outcome = response.Ok.outcome
  return 'Completed' in outcome || 'Failed' in outcome || 'ResultExpired' in outcome
    ? response.Ok
    : controlResult(await user.reconcile_execution(account, request))
}
